### ISSUE TITLE
Materialized view use at least 5 times more memory than when i'm doing the aggregation manually 

### ISSUE THREAD

**Issue by TH-HA** [ClickHouse/ClickHouse#6275]
Hello clickhouse team
I 'm trying to use a Materialized view with an aggregating mergetree to aggregate data automatically when they are inserted. 
The chain is very simple 
  insertion table is either a mergetree or a null engine ( both lead to the same result ) 
  The aggregating merge tree is define idependantly 
  and a MV  is added on the top to feed the aggregating merge tree with the result of an aggregation.
  
batch size are large ( 500 000 thousand rows ) , the result of the aggregation is very small: 280 rows. The table is pretty large ( 320 columns ).
The aggregation is also pretty simple mostly sum with a group by with access to a dictionary.

My issue is the following : 
   when i'm doing thing manually 
        insert into insertion table then do an insert into <aggregation table> select <aggregation request> 
         The queries use less than 1GB of memory.
   When i'm using the materialized view to do the same thing 
          I'm reaching memory limit ( setup to 10GB / default value ) 
	
So basically , what i'm seeing is a memory consumption at least 5 times higher when i'm using materialized view.

Is there any parameters to reduce this memory usage ? 
How can i troublehsoot this as clickhouse is hidding MV mechanism and we cannot see the split between initial insertion request , aggregating request and insertion in the aggregating merge tree.

I'm using version 19.11.4.2 ( but the problem was also visibile in 19.5.X) with a single node. 

Thanks in advance

**Comment by den-crane** [comment]
try to insert with max_block_size=100000 (with MV) (as a test)

**Comment by TH-HA** [comment]
@den-crane  , thanks for the parameter , i've  rerun the test with the parameter --max_insert_block_size=100000 add to the clickhouse-client. It's properly taken into account ( i can see the correct value in query_log but it does not change anything - same memory usage. 

Regards

**Comment by den-crane** [comment]
Can it be that user `default` and your current user have different max_bytes_before_external_group_by settings?

Check and share output of your insert with send_logs_level = 'debig'

**Comment by TH-HA** [comment]
i'm using the default user in all the case and i've tried to change the settings in multiple way 
thru the users.xml / and i can see the changed setting properly set  : but same result 
thru the clickhouse-client : same result 

The log are the following 
>


2019.08.07 17:10:05.953491 [ 42 ] {} <Trace> TCPHandlerFactory: TCP Request. Address: [::ffff:127.0.0.1]:40480
2019.08.07 17:10:05.953802 [ 42 ] {} <Debug> TCPHandler: Connected ClickHouse client version 19.11.0, revision: 54423, database: RE, user: default.
2019.08.07 17:10:05.974051 [ 42 ] {f4561fb4-1efa-4bd4-9251-df8219f995a6} <Debug> executeQuery: (from [::ffff:127.0.0.1]:40480) INSERT INTO INSERT_SUBSCRIBER FORMAT CSVWithNames
2019.08.07 17:10:05.980690 [ 18 ] {} <Trace> SystemLog (system.query_log): Flushing system log
2019.08.07 17:10:05.987058 [ 18 ] {} <Trace> system.query_log: Renaming temporary part tmp_insert_201908_8_8_0 to 201908_88007_88007_0.
2019.08.07 17:10:09.312259 [ 42 ] {f4561fb4-1efa-4bd4-9251-df8219f995a6} <Trace> InterpreterSelectQuery: FetchColumns -> Complete
2019.08.07 17:10:09.379332 [ 42 ] {f4561fb4-1efa-4bd4-9251-df8219f995a6} <Trace> Aggregator: Aggregating
2019.08.07 17:10:10.437088 [ 42 ] {f4561fb4-1efa-4bd4-9251-df8219f995a6} <Debug> MemoryTracker: Current memory usage: 1.06 GiB.
2019.08.07 17:10:11.807223 [ 42 ] {f4561fb4-1efa-4bd4-9251-df8219f995a6} <Debug> MemoryTracker: Current memory usage: 2.06 GiB.
2019.08.07 17:10:12.937203 [ 42 ] {f4561fb4-1efa-4bd4-9251-df8219f995a6} <Debug> MemoryTracker: Current memory usage: 3.06 GiB.
2019.08.07 17:10:14.021764 [ 42 ] {f4561fb4-1efa-4bd4-9251-df8219f995a6} <Debug> MemoryTracker: Current memory usage: 4.06 GiB.
2019.08.07 17:10:15.114920 [ 42 ] {f4561fb4-1efa-4bd4-9251-df8219f995a6} <Debug> MemoryTracker: Current memory usage: 5.01 GiB.
2019.08.07 17:10:16.237859 [ 42 ] {f4561fb4-1efa-4bd4-9251-df8219f995a6} <Debug> MemoryTracker: Current memory usage: 6.07 GiB.
2019.08.07 17:10:17.457666 [ 42 ] {f4561fb4-1efa-4bd4-9251-df8219f995a6} <Debug> MemoryTracker: Current memory usage: 7.07 GiB.
2019.08.07 17:10:18.784607 [ 42 ] {f4561fb4-1efa-4bd4-9251-df8219f995a6} <Debug> MemoryTracker: Current memory usage: 8.01 GiB.
2019.08.07 17:10:20.354915 [ 42 ] {f4561fb4-1efa-4bd4-9251-df8219f995a6} <Debug> MemoryTracker: Current memory usage: 9.01 GiB.
2019.08.07 17:10:21.492788 [ 42 ] {f4561fb4-1efa-4bd4-9251-df8219f995a6} <Error> executeQuery: Code: 241, e.displayText() = DB::Exception: Memory limit (for query) exceeded: would use 9.32 GiB (attempt to allocate chunk of 67108864 bytes), maximum: 9.31 GiB: while pushing to view RE.MV_SUBSCRIBERGROUP_AGGSUBSCRIBERGROUP (version 19.11.4.24 (official build)) (from [::ffff:127.0.0.1]:40480) (in query: INSERT INTO INSERT_SUBSCRIBER FORMAT CSVWithNames), Stack trace:

0. /usr/bin/clickhouse-server(StackTrace::StackTrace()+0x30) [0x7d2c950]
1. /usr/bin/clickhouse-server(DB::Exception::Exception(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, int)+0x25) [0x3bb8975]
2. /usr/bin/clickhouse-server() [0x3a2748a]
3. /usr/bin/clickhouse-server(MemoryTracker::alloc(long)+0xc4) [0x767daa4]
4. /usr/bin/clickhouse-server(AllocatorWithHint<false, AllocatorHints::DefaultHint, 67108864ul>::alloc(unsigned long, unsigned long)+0x31) [0x3bc5a31]
5. /usr/bin/clickhouse-server() [0x6dab6ca]
6. /usr/bin/clickhouse-server(DB::ColumnVector<unsigned int>::replicate(DB::PODArray<unsigned long, 4096ul, AllocatorWithHint<false, AllocatorHints::DefaultHint, 67108864ul>, 15ul, 16ul> const&) const+0x339) [0x6dc5a39]
7. /usr/bin/clickhouse-server(DB::ExpressionAction::execute(DB::Block&, bool) const+0xa71) [0x723e711]
8. /usr/bin/clickhouse-server(DB::ExpressionActions::execute(DB::Block&, bool) const+0x45) [0x723f645]
9. /usr/bin/clickhouse-server(DB::FilterBlockInputStream::readImpl()+0x408) [0x70e2b28]
10. /usr/bin/clickhouse-server(DB::IBlockInputStream::read()+0x187) [0x6adbb77]
11. /usr/bin/clickhouse-server(DB::ExpressionBlockInputStream::readImpl()+0x2f) [0x70e175f]
12. /usr/bin/clickhouse-server(DB::IBlockInputStream::read()+0x187) [0x6adbb77]
13. /usr/bin/clickhouse-server(DB::ConcatBlockInputStream::readImpl()+0xbf) [0x6c3666f]
14. /usr/bin/clickhouse-server(DB::IBlockInputStream::read()+0x187) [0x6adbb77]
15. /usr/bin/clickhouse-server(DB::Aggregator::execute(std::shared_ptr<DB::IBlockInputStream> const&, DB::AggregatedDataVariants&)+0x383) [0x7162c33]
16. /usr/bin/clickhouse-server(DB::AggregatingBlockInputStream::readImpl()+0x432) [0x70acdb2]
17. /usr/bin/clickhouse-server(DB::IBlockInputStream::read()+0x187) [0x6adbb77]
18. /usr/bin/clickhouse-server(DB::ExpressionBlockInputStream::readImpl()+0x2f) [0x70e175f]
19. /usr/bin/clickhouse-server(DB::IBlockInputStream::read()+0x187) [0x6adbb77]
20. /usr/bin/clickhouse-server(DB::ExpressionBlockInputStream::readImpl()+0x2f) [0x70e175f]
21. /usr/bin/clickhouse-server(DB::IBlockInputStream::read()+0x187) [0x6adbb77]
22. /usr/bin/clickhouse-server(DB::MaterializingBlockInputStream::readImpl()+0x3a) [0x70eabba]
23. /usr/bin/clickhouse-server(DB::IBlockInputStream::read()+0x187) [0x6adbb77]
24. /usr/bin/clickhouse-server(DB::SquashingBlockInputStream::readImpl()+0x86) [0x6af2396]
25. /usr/bin/clickhouse-server(DB::IBlockInputStream::read()+0x187) [0x6adbb77]
26. /usr/bin/clickhouse-server(DB::PushingToViewsBlockOutputStream::process(DB::Block const&, unsigned long)+0x82f) [0x711478f]
27. /usr/bin/clickhouse-server(DB::PushingToViewsBlockOutputStream::write(DB::Block const&)+0x322) [0x7115142]
28. /usr/bin/clickhouse-server(DB::SquashingBlockOutputStream::write(DB::Block const&)+0x482) [0x711dcf2]
29. /usr/bin/clickhouse-server(DB::AddingDefaultBlockOutputStream::write(DB::Block const&)+0x8f) [0x70abb2f]
30. /usr/bin/clickhouse-server(DB::CountingBlockOutputStream::write(DB::Block const&)+0x2c) [0x70c62fc]
31. /usr/bin/clickhouse-server(DB::TCPHandler::receiveData()+0xb0) [0x3bc15a0]

2019.08.07 17:10:21.492863 [ 19 ] {} <Trace> SystemLog (system.query_thread_log): Flushing system log
2019.08.07 17:10:21.493099 [ 18 ] {} <Trace> SystemLog (system.query_log): Flushing system log
2019.08.07 17:10:21.493192 [ 42 ] {f4561fb4-1efa-4bd4-9251-df8219f995a6} <Debug> MemoryTracker: Peak memory usage (for query): 9.26 GiB.
2019.08.07 17:10:21.493937 [ 42 ] {f4561fb4-1efa-4bd4-9251-df8219f995a6} <Debug> MemoryTracker: Peak memory usage (total): 9.26 GiB.
2019.08.07 17:10:21.494006 [ 42 ] {f4561fb4-1efa-4bd4-9251-df8219f995a6} <Information> TCPHandler: Processed in 15.520 sec.
2019.08.07 17:10:21.494568 [ 42 ] {f4561fb4-1efa-4bd4-9251-df8219f995a6} <Information> TCPHandler: Processed in 0.000 sec.
2019.08.07 17:10:21.494671 [ 42 ] {f4561fb4-1efa-4bd4-9251-df8219f995a6} <Information> TCPHandler: Processed in 0.000 sec.
2019.08.07 17:10:21.504194 [ 19 ] {} <Trace> system.query_thread_log: Renaming temporary part tmp_insert_201908_6_6_0 to 201908_89127_89127_0.
2019.08.07 17:10:21.532706 [ 18 ] {} <Trace> system.query_log: Renaming temporary part tmp_insert_201908_9_9_0 to 201908_88008_88008_0.
2019.08.07 17:10:21.546568 [ 42 ] {} <Error> ServerErrorHandler: Code: 99, e.displayText() = DB::Exception: Unknown packet 1363637 from client, Stack trace:

0. /usr/bin/clickhouse-server(StackTrace::StackTrace()+0x30) [0x7d2c950]
1. /usr/bin/clickhouse-server(DB::Exception::Exception(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, int)+0x25) [0x3bb8975]
2. /usr/bin/clickhouse-server() [0x3757642]
3. /usr/bin/clickhouse-server(DB::TCPHandler::runImpl()+0x728) [0x3bc3028]
4. /usr/bin/clickhouse-server(DB::TCPHandler::run()+0x2b) [0x3bc383b]
5. /usr/bin/clickhouse-server(Poco::Net::TCPServerConnection::start()+0x10) [0x77077d0]
6. /usr/bin/clickhouse-server(Poco::Net::TCPServerDispatcher::run()+0xed) [0x7707eed]
7. /usr/bin/clickhouse-server(Poco::PooledThread::run()+0x81) [0x7e2b6a1]
8. /usr/bin/clickhouse-server(Poco::ThreadImpl::runnableEntry(void*)+0x3c) [0x7e2944c]
9. /usr/bin/clickhouse-server() [0xb7fec00]
10. /lib/x86_64-linux-gnu/libpthread.so.0(+0x76ba) [0x7f50654c46ba]
11. /lib/x86_64-linux-gnu/libc.so.6(clone+0x6d) [0x7f5064ced41d]
 (version 19.11.4.24 (official build))


I've also tried the following approach: 
   map the input csv as a table using file(....) and do an insert as select. Memory consumption does not exceed 3GB in this case.

**Comment by bianxintong** [comment]
can we have more details on the cause and progress on possible fixes? we're experiencing same issue.

**Comment by blinkov** [comment]
@TH-HA, do you have any further questions?

**Comment by melvynator** [comment]
Did you try lowering the settings: https://clickhouse.com/docs/en/operations/settings/settings#min-insert-block-size-rows-for-materialized-views ?
